text/0149-exponential-histogram.md

```diff
@@ -0,0 +1,162 @@
+
+## Explanation
+
+Exponential buckets will be added. In general, bucket bounds are in the form of:
+
+```
+bound = base ^ exponent
+```
+
+where base is a parameter of a bound series, and exponent is an integer. Note that exponent may be positive, negative, or 0. Such bounds are also commonly known as "log scale bounds".
+
+## Internal details
+
+Proposed message type to be added:
+
+```
+message ExponentialBuckets {
+    double base = 1;
+    double zero_count = 2; // Count of values exactly at zero.
+    ExponentialBucketCounts positive_value_counts = 3;
+    ExponentialBucketCounts negative_value_counts = 4; // Negative values are bucketed with their absolute values
+}
+
+// "repeated double bucket_counts" represents an array of N numbers from bucket_counts[0] to bucket_counts[N-1].
+// With index i starting at 0, ending at N-1, ExponentialBucketCounts defines N buckets, where
+// bucket[i].start_bound = base ^ (i + exponent_offset)
+// bucket[i].end_bound = base ^ (i + 1 + exponent_offset)
+// bucket[i].count = bucket_counts[i]
+message ExponentialBucketCounts {
+    sint32 exponent_offset = 1; // offset may be negative.
+    repeated double bucket_counts = 2;
+}
+```
+
+Notes:
+
+* ExponentialBuckets will be added as "oneof" the bucket types in [#272](https://github.com/open-telemetry/opentelemetry-proto/pull/272)
+* Per [#257](https://github.com/open-telemetry/opentelemetry-proto/issues/257), only a histogram accepting "double" will be defined.
+* Per [#259](https://github.com/open-telemetry/opentelemetry-proto/issues/259), bucket counts type is "double".
+
+## Trade-offs and mitigations
+
+Simplicity is a main design goal. The format targets the most common scenarios. For now, histograms not conforming to ExponentialBuckets may be encoded as explicit buckets. If a histogram type is common enough, a new bucket type may be added in the future.
+
+The followings are restrictions of ExponentialBuckets:
+
+* Buckets for positive and negative values must have the same "base".
+* Buckets must cover the full value range. In the future, ExponentialBucketCounts might add an overflow_count and an underflow_count for counts above the highest bucket and below the lowest buckets, respectively. However, overflow or underflow bucket breaks the "pure log scale" property. "Rescale" is preferred when reducing memory cost. See "merge" section below.
+* ExponentialBucketCounts is designed for dense buckets. Between the lowest and the highest bucket, most buckets are expected to be non-empty. This is the common case in telemetry applications.
```

Comment:
> ExponentialBucketCounts is designed for dense buckets. Between the lowest and the highest bucket, most buckets are expected to be non-empty. This is the common case in telemetry applications.

My experience is somewhat different here:

a) Latency metrics have a long tail, if your bucketing scheme has 1% accuracy, there might be many empty buckets before you reach the latency outliers (>10sec).
b) In order to avoid up-front choices it's very convenient to have a lot's of headroom at the top and bottom end of your scale, e.g. covering 10^+/-100, in this case you will have lot's of empty buckets as well. The problem is very similar to floating point arithmetic: When you declare a float you don't want to provide min-max exponents.

The conclusion that we (back at Circonus [1]) and now also Prometheus [2] reached is, that sparse encoding is the better alternative.

[1] [Circllhist](https://arxiv.org/abs/2001.06561)
[2] [Sparse high-resolution histograms for Prometheus](https://docs.google.com/document/d/1cLNv3aufPZb3fNfaJgdaRBZsInZKKIHo9E6HinJVbpM/edit#)

Reply:
There are many known methods for sparse array encoding. I'd like to start with a simple array and extend to fancier forms if necessary. 
Sparse vs. dense may have to do with config. In New Relic "Distribution" metric, the number of buckets is limited to 320. A typical histogram aggregates over thousands to millions of values, which are likely to make the buckets dense.  
More discussion in the "config" comment below. 
For "b)", many implementation, including DDSketch and New Relic (internal use) dynamically shift the array to cover only the dense part, avoiding empty buckets at the bottom and top.

Reply:
This makes sense. I can see a dynamically allocated (!) dense-region working in practice.